app/models/match.rb

```diff
@@ -7,5 +7,6 @@ class Match < ApplicationRecord
   has_many :users, through: :predictions
   validates :kickoff_time, presence: true
   validates :status, presence: true
-  enum status: %i[awaiting started finished]
+  enum status: %i[upcoming started finished]
+  validates_uniqueness_of :kickoff_time, scope: [:team_home, :team_away]
```

Comment:
kickoff_time is a DateTime so cant be two teams at same time and date.

but i do think we might have an issue for the knockout rounds now that we removed round_id from a match. unless actually we create groups for the knockout stages

Reply:
Makes sense, forgot the date was in there too 🤦‍♂️

Good point for the knockout rounds. Creating a group for each round of the knockouts might be a bit hacky... We could reintroduce a `round_id` in the matches, but add a validation so that only either `group_id` or `round_id` can be filled at the same time (to avoid potential conflicts). We can also override the `Match#round` method so that it returns automatically the right round via `round_id` or `group_id`, whichever is filled.

Reply:
or make it polymorphic?

Reply:
Oh cool sexy words... :)

Haven't got much experience using polymorphic associations, but our Group and Round models are quite different. If we make it polymorphic, we'd probably still need to check which one we have when using them, which would defeat the purpose of the polymorphism, no? In our case having access to group or round separately might be more practical?

Typical example would be if we want to retrieve the round number, we would first need to check if the polymorphism returns a round or not. If not we'd need to do `polymorphism.round` which also seems kinda wrong, no?

Again, not much experience there, so let me know if I'm having the wrong idea of how this would work.

Reply:
yeah that's a fair point. i think in polymorphic, you're usually checking anyway. However, like you said, we'll be using them in different way so might just be easier to have both.

Reply:
Yeah my assumption was always that you'd try to use polymorphism paired with duck typing? Like restaurants, bars and nightclubs for example, which would share many common attributes/methods.

The closest practical example I have in mind would be the Movies and TV Series during the search lecture. They share most attributes except for the director, so you can still pretty much iterate over all of them without having to do much checking.

But in any case, I never have the reflex of thinking about polymorphisms, I should add this to my toolbox.

Reply:
Yeah I used it a couple times at my last job but would have to look back to see when the times were appropriate. The only time i've used it since is when i made the Le Wagon projects app. I either connected a `team` or a `person` to a `project`